proto/asn1-pdu-proto/asn1_proto_to_der.cc

```diff
@@ -0,0 +1,158 @@
+  } else {
+    EncodeDefiniteLength(actual_len, len_pos);
+  }
+}
+
+void ASN1ProtoToDER::EncodeValue(const Value& val) {
+  for (const auto& val_ele : val.val_array()) {
+    if (!kErrRecursionLimitReached) {
+      if (val_ele.has_pdu()) {
+        EncodePDU(val_ele.pdu());
+      } else {
+        encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
+                        val_ele.val_bits().end());
+      }
+    }
+  }
+}
+
+void ASN1ProtoToDER::EncodeHighTagNumberForm(const uint8_t id_class,
+                                             const uint8_t encoding,
+                                             const uint32_t tag_num) {
+  // The high-tag-number form base 128 encodes |tag_num| (X.690 (2015), 8.1.2).
+  uint8_t num_bytes = GetVariableIntLen(tag_num, 128);
+  // High-tag-number form requires the lower 5 bits of the identifier to be set
+  // to 1 (X.690 (2015), 8.1.2.4.1).
+  uint64_t id_parsed = (id_class | encoding | 0x1F);
+  id_parsed <<= 8;
+  for (uint8_t i = num_bytes - 1; i != 0; --i) {
+    // If it's not the last byte, the high bit is set to 1 (X.690
+    // (2015), 8.1.2.4.2).
+    id_parsed |= ((0x01 << 7) | ((tag_num >> (i * 7)) & 0x7F));
+    id_parsed <<= 8;
+  }
+  id_parsed |= (tag_num & 0x7F);
+  InsertVariableInt(id_parsed, encoder_.size());
+}
+
+void ASN1ProtoToDER::EncodeIdentifier(const Identifier& id) {
+  // The class comprises the 7th and 8th bit of the identifier (X.690
+  // (2015), 8.1.2).
+  uint8_t id_class = static_cast<uint8_t>(id.id_class()) << 6;
+  // The encoding comprises the 6th bit of the identifier (X.690 (2015), 8.1.2).
+  uint8_t encoding = static_cast<uint8_t>(id.encoding()) << 5;
+
+  uint32_t tag_num = id.tag_num().has_high_tag_num()
+                         ? id.tag_num().high_tag_num()
+                         : id.tag_num().low_tag_num();
+  // When the tag number is greater than or equal to 31, encode with a single
+  // byte; otherwise, use the high-tag-number form (X.690 (2015), 8.1.2).
+  if (tag_num >= 31) {
+    EncodeHighTagNumberForm(id_class, encoding, tag_num);
+  } else {
+    encoder_.push_back(static_cast<uint8_t>(id_class | encoding | tag_num));
+  }
+}
+
+void ASN1ProtoToDER::EncodePDU(const PDU& pdu) {
+  ++depth_;
+  // Artifically limit the stack depth to avoid stack overflow.
+  if (depth_ > kRecursionLimit) {
+    kErrRecursionLimitReached = true;
+    return;
+  }
+  size_t size_before_insertion = encoder_.size();
+  EncodeIdentifier(pdu.id());
+  size_t id_size = encoder_.size() - size_before_insertion;
+  EncodeValue(pdu.val());
+  size_t value_size = encoder_.size() - size_before_insertion - id_size;
+  EncodeLength(pdu.len(), value_size, size_before_insertion + id_size);
+  --depth_;
+}
+
+std::vector<uint8_t> ASN1ProtoToDER::ProtoToDER(const PDU& pdu) {
+  // The same instantiaton of ANS1ProtoToDER can encode multiple PDU's.
+  // The following three instructions are used to reset the values
+  // before each PDU is encoded.
```

Comment:
I'd say this is fairly obvious, and so probably doesn't need a comment, although if you do comment, probably something as simple as.

```suggestion
  // Reset the previous state.
```